File: src/types/triage.ts
export type TriageRequest = {
  symptoms: string
  clinical_context?: string
  image?: File | Blob
}

export type TriageResponse = {
  assessment_id: string
  urgency_level: string
  rationale: string
  possible_conditions: string[]
  recommended_actions: string[]
  disclaimer: string
}

export type TriageChatRequest = {
  symptoms: string
  messages?: string
  /** JSON string: profile details, marked body areas, self-rated severity */
  context?: string
  image?: File | Blob
}

export type CarePlan = {
  immediate_relief: string[]
  food_and_water: string[]
  when_to_hospital: string[]
}

/** The five levels from the AI Engineer Handoff. The app, not the model, decides these. */
export type Urgency = 'EMERGENCY' | 'URGENT' | 'SOON' | 'SELF_CARE' | 'INSUFFICIENT_INFORMATION'

export type TriageIndicator = {
  label: string
  color: 'red' | 'orange' | 'yellow' | 'green' | 'gray'
  icon: string
  priority: number
}

export type FollowUpQuestion = {
  id: string
  text: string
  type: 'single_choice' | 'short_text'
  options: string[]
  required: boolean
}

export type TriageChatResponse = {
  assessment_id: string
  needs_more_info: boolean
  urgency_level: string
  /** No longer produced by the server (the old values were fixed numbers). */
  confidence_score: number | null
  rationale: string
  possible_conditions: string[]
  /** Locked 3-part care plan: immediate relief, food & water, when to go to hospital. */
  care_plan?: CarePlan
  /** Kept for older sessions: recommended_actions/red_flags_to_watch mirror care_plan. */
  recommended_actions: string[]
  follow_up_questions: string[]
  red_flags_to_watch: string[]
  disclaimer: string
  reply?: string
  has_symptoms?: boolean
  is_conversational?: boolean
  /** "gemini" = real AI answer, "rules" = server safety rules only, "offline" = local demo fallback */
  ai_source?: 'gemini' | 'rules' | 'offline'
  ai_notice?: string
  /** Things the assistant just remembered about the user */
  memory_notes?: string[]

  // The handoff result contract. Present on every new answer; older saved sessions don't have it.
  status?: 'question' | 'complete' | 'emergency_stop'
  urgency?: Urgency
  indicator?: TriageIndicator
  summary?: string
  reason?: string | null
  next_steps?: string[]
  escalation?: { required: boolean; message: string | null }
  facility_action?: string | null
  follow_up_question?: FollowUpQuestion | null
  safety_note?: string
  red_flags?: string[]
  warning_signs?: string[]
  rule_version?: string
  medication_notice?: string
  /** Approved-flow state: sent back unchanged on the next message. */
  flow?: Record<string, unknown> | null
  /** The age profile the assessment used (addendum section 4). */
  profile?: { band: AgeBand; label: string; age: string | null; weight_kg: number | null; for: PatientFor } | null
  /** Which approved table was used (typhoid, respiratory, hypertension, diarrhea, under_6), if any. */
  pathway?: string | null
}

export type AgeBand = 'adult' | 'pediatric_6_plus' | 'under_6'
export type PatientFor = 'self' | 'child' | 'other'

/** Who the check is for. Weight is recorded for the future dosage module; triage does not use it. */
export type PatientInfo = {
  for: PatientFor
  age_years?: number | null
  age_months?: number | null
  weight_kg?: number | null
  pregnant?: 'yes' | 'no' | 'not_sure' | null
}

export type CacheStats = {
  hits: number
  misses: number
  hit_rate: number
  size: number
}

export type CacheClearResponse = {
  status: string
}

export type AiPreferences = {
  response_style: 'concise' | 'balanced' | 'detailed'
  tone: 'gentle' | 'direct'
  units: 'metric' | 'imperial'
  language: string
  emergency_number: string
  remember_conversations: boolean
}

export type AiHealthContext = {
  age: number | null
  gender: string
  location: string
  conditions: string[]
  allergies: string[]
  medications: string[]
}

export type AiMemory = {
  preferences: AiPreferences
  health_context: AiHealthContext
  facts: { id: string; text: string; source: string; created_at: string }[]
  history: { at: string; field: string; from: unknown; to: unknown; source: 'user' | 'ai' | 'profile' }[]
  updated_at: string | null
}

export type AiKeyInfo = {
  id: string
  label: string
  masked: string
  model: string
  enabled: boolean
  source: 'env' | 'admin'
  added_at: string | null
  uses: number
  failures: number
  last_used: string | null
  last_error: string
  status: 'unknown' | 'ok' | 'rate_limited' | 'invalid' | 'error' | 'cooling_down'
  cooldown_seconds: number
}

export type AiStatus = { ai_enabled: boolean; keys_total: number; keys_active: number }

File: backend/app/schemas/triage.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class TriageRequest(BaseModel):
    symptoms: str
    clinical_context: Optional[str] = None


class TriageResponse(BaseModel):
    assessment_id: str
    urgency_level: str
    rationale: str
    possible_conditions: List[str]
    recommended_actions: List[str]
    disclaimer: str
    ai_source: str = "rules"
    ai_notice: str = ""
    # The AI Engineer Handoff result contract (section 4). The fields above are kept
    # for the current frontend and are derived from these.
    status: Optional[str] = None
    urgency: Optional[str] = None
    indicator: Optional[Dict[str, Any]] = None
    summary: Optional[str] = None
    reason: Optional[str] = None
    next_steps: List[str] = []
    escalation: Optional[Dict[str, Any]] = None
    facility_action: Optional[str] = None
    follow_up_question: Optional[Dict[str, Any]] = None
    safety_note: Optional[str] = None
    red_flags: List[str] = []
    warning_signs: List[str] = []
    rule_version: Optional[str] = None
    medication_notice: str = ""
    # Approved-flow state (sent back by the client) and the age profile used (addendum 4).
    flow: Optional[Dict[str, Any]] = None
    profile: Optional[Dict[str, Any]] = None
    pathway: Optional[str] = None


class TriageChatRequest(BaseModel):
    symptoms: str
    messages: Optional[str] = "[]"


class TriageChatResponse(BaseModel):
    assessment_id: str
    needs_more_info: bool
    urgency_level: str
    # No longer produced: the old values were fixed numbers, not a measured confidence.
    confidence_score: Optional[float] = None
    rationale: str
    possible_conditions: List[str]
    recommended_actions: List[str]
    follow_up_questions: List[str]
    red_flags_to_watch: List[str]
    disclaimer: str
    reply: Optional[str] = ""
    has_symptoms: Optional[bool] = False
    is_conversational: Optional[bool] = False
    ai_source: str = "rules"  # "gemini" | "rules"
    ai_notice: str = ""
    memory_notes: List[str] = []
    # The AI Engineer Handoff result contract (section 4). The fields above are kept
    # for the current frontend and are derived from these.
    status: Optional[str] = None
    urgency: Optional[str] = None
    indicator: Optional[Dict[str, Any]] = None
    summary: Optional[str] = None
    reason: Optional[str] = None
    next_steps: List[str] = []
    escalation: Optional[Dict[str, Any]] = None
    facility_action: Optional[str] = None
    follow_up_question: Optional[Dict[str, Any]] = None
    safety_note: Optional[str] = None
    red_flags: List[str] = []
    warning_signs: List[str] = []
    rule_version: Optional[str] = None
    medication_notice: str = ""
    # Approved-flow state (sent back by the client) and the age profile used (addendum 4).
    flow: Optional[Dict[str, Any]] = None
    profile: Optional[Dict[str, Any]] = None
    pathway: Optional[str] = None


class BackendTriageStatus(BaseModel):
    level: str  # 'Emergency' | 'Urgent' | 'Non-Urgent'


class BackendTriageItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    # Pydantic v2 treats a leading-underscore name as private and never serialises it, so alias it.
    id: str = Field(alias="_id")
    symptoms: List[str]
    duration: str = "Recent"
    severity: str = "Moderate"  # 'Mild' | 'Moderate' | 'Severe'
    notes: Optional[str] = ""
    triageStatus: BackendTriageStatus
    actionPlan: str
    createdAt: str
    possible_conditions: Optional[List[str]] = []
    recommended_actions: Optional[List[str]] = []
    urgency_level: Optional[str] = None
    rationale: Optional[str] = ""


class TriageListResponse(BaseModel):
    msg: str = "Triage history retrieved"
    data: List[BackendTriageItem]


class CacheStats(BaseModel):
    hits: int
    misses: int
    hit_rate: float
    size: int


class CacheClearResponse(BaseModel):
    status: str
